src/processor.cpp

```diff
@@ -0,0 +1,401 @@
+        std::vector<item> materials;
+        std::vector<integer> charges;
+        item &fuel;
+        items = g->m.i_at( examp );
+        for( auto item_it = items.begin(); item_it != items.end(); ) {
+            if (item_it.typeId() == processor.fuel_type ){
+                fuel = item_it;
+                continue;
+            }
+            for( auto component_it = processed_materials.begin(); component_it != processed_materials.end(); ){
+                //If we found which components it belongs to
+                if( item_it.typeId() == component_it.typeId() ) {
+                    materials.push_back( item_it.typeId() );
+                    charges.push_back( item_it.charges );
+                    break;
+                }
+                component_it++;
+            }
+            item_it++;
+        }
+        //  Now we know how many charges of each components do we have, along with how much fuel
+        //  Current implementation plan is to assume use of one on each components and one fuel_intake number of fuel charges
+        //  per charge of output produced
+
+        //  Should it be volume based?
+        //TODO: calculate skill gain
+        //TODO: calculate output here, or in second phase, based on charges and fuel present at start
+
+        //TODO: make it skill use dependent? Would be silly to ask for fire on a CNC machine down the road
+        if( !p.has_charges( "fire" , 1 ) ) {
+            add_msg( _("This process is ready to be started, but you have no fire source.") );
+            return;
+        } else if( !query_yn( _( processor.full_message ) ) ) {
+            return;
+        }
+        add_msg( _( processor.start_message ) );
+        p.use_charges( "fire", 1 );
+        //TODO: skill xp gain
+
+        for( auto item_it = items.begin(); item_it != items.end(); ){
+            item_it.bday = calendar::turn;
+        }
+        g->m.furn_set(examp, processor.next_type );
+    } else {
+        //processor processing away
+        auto items = g->m.i_at( examp );
+        if( items.empty() ) {
+            debugmsg( processor.empty_message );
+            g->m.furn_set( examp, processor.next_type );
+            return;
+        }
+        int last_bday = items[0].bday;
+        for( auto i : items ) {
+            if( processed_materials.find( item_it->id ) != container.end() && i.bday > last_bday ) {
+                last_bday = i.bday;
+            }
+        }
+        int process_time = active_process.duration;
+        int time_left = process_time - calendar::turn.get_turn() + items[0].bday;
+        if( time_left > 0 ) {
+            add_msg( _("It should take %d minutes to finish."), time_left / MINUTES(1) + 1 );
+            return;
+        }
+
+        //  Currently in a very unordered fashion (keg code could help)
+        //  This next loop is currently not used, could help sort items into a
+        //  predefined order (based on order in process.json maybe?)
```

Comment:
You want to use `std::map<itype_id, int>` here.
So you can call something like `components["clay"] += item_here.charges`.